### PR TITLE
library: add Drop Zone Entry

### LINKED ISSUE #368: Drop Zone entry
We should have a demo that demonstrates how to have a drop zone which can be used to drop a file or text or anything from outside the app.

I would suggest making something like this to start with

EDIT: nevermind this design refer to the commet below

image

The zone should highlight or animate when there is a drag over it. 
In GNOME we usually show an inset border of 1px.  See for example with https://apps.gnome.org/app/app.drey.Warp/
There is also a good example here: https://www.dropzone.dev/

Code:

```blp
using Gtk 4.0;
using Adw 1;

Adw.StatusPage {
  title: "Drop Zone";
  description: "Accept data dragged from other applications";
  Box {
    margin-top: 48;
    hexpand: true;
    vexpand: true;
    halign: center;
    valign: center;
    orientation: vertical;

    Box {
      name: "drop";
      width-request: 200;
      height-request: 200;
      Label {
        name: "label";
        label: _("<b>Drop</b>");
        use-markup: true;
        hexpand: true;
      }
    }
  }
}

```

```css
#drop {
  border: dashed 6px @dark_3;
  border-radius: 25%;
  background-color: white;
  
}

#label {
  color: @dark_3;
}
```

### PR DESCRIPTION
**Pull request by halfmexican**
drafting this 

closes #368

### CODE REVIEW

File: src/Library/demos/Drop Zone/main.blp [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  title: _("Drop Zone");
+  description: _("A versatile drop area for external content");
+
+  Box {
+    orientation: vertical;
+
+    Adw.Clamp {
+      maximum-size: 500;
+
+      Adw.Bin bin {
```
Comment (sonnyp) [inline src/Library/demos/Drop Zone/main.blp:14]: `Bin` isn't really meant to be used directly but rather for subclassing

Would using `Box` complicate the code significantly?

File: src/Library/demos/Drop Zone/main.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  if (!(value instanceof Gio.File)) return false;
+
+  const file_info = value.query_info("standard::content-type", 0, null);
+  const content_type = file_info.get_content_type();
+
+  if (content_type.startsWith("image/")) {
+    try {
+      bin.child = createImageWidget(value);
+    } catch (error) {
+      console.log(`Unable to load image: ${error}`);
```
Comment (sonnyp) [inline src/Library/demos/Drop Zone/main.js:36]: always use `logError` in GJS (at least for now, until we fix `console` to deal with it properly)

File: src/Library/demos/Drop Zone/main.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  } else {
+    try {
+      bin.child = createFileWidget(value);
+    } catch (error) {
+      console.log(`Unable to create file widget: ${error}`);
+    }
+  }
+
+  bin.get_style_context().remove_class("overlay-drag-area");
+});
```
Comment (sonnyp) [inline src/Library/demos/Drop Zone/main.js:53]: you can simplify this by moving the content of the signal handler to a new function `onDrop` 

and moving the try/catch upwards:

```
file_drop_target.connect("drop", (self, value, x, y) => {
  try {
    onDrop(value);
  } catch (err) {
    logError(err);
  }
});
```

File: src/Library/demos/Drop Zone/main.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  widget.append(icon);
+  icon.icon_size = Gtk.IconSize.LARGE;
+
+  const file_name = new Gtk.Label({ label: file.get_basename() });
+  widget.append(file_name);
+
+  return widget;
+}
+
+function createBoxWidget() {
```
Comment (sonnyp) [inline src/Library/demos/Drop Zone/main.js:95]: FYI, you could move this to Blueprint but it's fine here too

File: src/Library/demos/Drop Zone/main.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+const string_drop_target = Gtk.DropTarget.new(
+  GObject.TYPE_STRING,
+  Gdk.DragAction.COPY,
+);
+
+bin.add_controller(string_drop_target);
+
+string_drop_target.connect("drop", (self, value, x, y) => {
+  bin.child = createTextWidget(value);
+  bin.get_style_context().remove_class("overlay-drag-area");
```
Comment (sonnyp) [inline src/Library/demos/Drop Zone/main.js:19]: `get_style_context` is deprecated.

Use `add_css_class` and `remove_css_class` Widget methods directly

File: src/Library/demos/Drop Zone/main.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+function createVideoWidget(file) {
+  const widget = createBoxWidget();
+
+  const video = new Gtk.Video({ file: file });
+  widget.append(video);
+  return widget;
+}
+
+function createFileWidget(file) {
```
Comment (sonnyp) [inline src/Library/demos/Drop Zone/main.js:81]: maybe rename the functions to `createFooPreview` ?

File: src/Library/demos/Drop Zone/main.blp [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+          label: _("Drop files or text here");
+        }
+      }
+    }
+
+    LinkButton {
+      label: "API Reference";
+      uri: "https://docs.gtk.org/gtk4/drag-and-drop.html";
+      margin-top: 24;
+    }
```
Comment (sonnyp) [inline src/Library/demos/Drop Zone/main.blp:28]: this is not an API reference, here is a suggestion

```
    Box {
      margin-top: 24;
      orientation: vertical;

      LinkButton {
        label: "Drag-and-Drop in GTK";
        uri: "https://docs.gtk.org/gtk4/drag-and-drop.html#drop-targets";
      }
      
      LinkButton {
        label: "DropTarget API Reference";
        uri: "https://docs.gtk.org/gtk4/class.DropTarget.html";
      }
    }
```

**Review by sonnyp — CHANGES_REQUESTED** [review]
Excellent!

Just a couple more things for perfection:

Can you support dragging data too? You can test with dropping an image displayed in your web browser. Dragging text from the browser also does not work but I suspect it's the same problem.

If you look at the paper clip screenshot, you'll notice there is a margin around the blue overlay:

Screenshot from 2023-07-08 23-13-52

File: src/Library/demos/Drop Zone/main.blp (hunk 1 above)
Comment (halfmexican) [reply src/Library/demos/Drop Zone/main.blp:14]: It's much easier to use a bin as I can just set the child parameter, with a box i would have to handle the removing a child from the box whenever another item is dropped. The document page said that "it is useful for deriving subclasses" does that mean we have to?

Comment (sonnyp) [reply src/Library/demos/Drop Zone/main.blp]: Let's use it then :+1:

**Review by sonnyp — APPROVED** [review]
Great work, I created a follow up https://github.com/sonnyp/Workbench/issues/440

**Pull request merged**